### PR TITLE
New Project Script

### PR DESCRIPTION
**Pull request by Filmaluco**
added: newproject script
allows to start a new project with a given name, avoiding manual edition of files

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,6 +28,15 @@
 
 # Getting started
 
+### Script 
+1. Download this repository extract and open the folder
+2. Run `newproject.sh` script to create a copy with the name you want
```
Comment (sdsantos) [inline README.md:37]: You can provide here an example of a command running the newproject.sh.

File: README.md [hunk 2]
```diff
@@ -28,6 +28,15 @@
 
 # Getting started
 
+### Script 
+1. Download this repository extract and open the folder
+2. Run `newproject.sh` script to create a copy with the name you want 
+3. On `app`, change the package on the `mode` lazy var
```
Comment (sdsantos) [inline README.md:38]: No longer needed.

File: README.md [hunk 3]
```diff
@@ -28,6 +28,15 @@
...
+4. On `theme.xml` & `colors.xml` set your application primary and secondary colors 
+
+And you're ready to start working on your new app.
+
+### Manual
 1. Download this repository extract and open the template folder on Android Studio
 2. Rename the app package `io.bloco.template`
 3. Check if the manifest package was renamed along with the package
```
Comment (sdsantos) [inline README.md:46]: The step 4 is no longer needed.

Comment (sdsantos) [reply README.md:46]: We either add step 6 to the Script as well, or take it out.

Comment (sdsantos) [reply README.md:46]: `Make sure the method `checkTestMode` inside `AndroidApplication` includes a test class that exists` no longer applies

File: newproject.sh [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+# get rid of gradle cache
+rm -rf $TITLE_NO_SPACES/.gradle
+# get rid of the git history
+rm -rf $TITLE_NO_SPACES/.git
+# get rid of the build
+rm -rf $TITLE_NO_SPACES/build
+rm -rf $TITLE_NO_SPACES/app/build
+
+# Rename folder structure
+renamefolderstructure() {
```
Comment (sdsantos) [inline newproject.sh:97]: No camelCase available?

File: newproject.sh [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+# search and replace in files
+PACKAGE_NAME_ESCAPED="${packagename//./\.}"
+OLD_PACKAGE_NAME_ESCAPED="${OLD_PACKAGE//./\.}"
+LC_ALL=C find $WORKING_DIR/$TITLE_NO_SPACES -type f -exec sed -i "" "s/$OLD_PACKAGE_NAME_ESCAPED/$PACKAGE_NAME_ESCAPED/g" {} +
+LC_ALL=C find $WORKING_DIR/$TITLE_NO_SPACES -type f -exec sed -i "" "s/$OLD_TITLE/$TITLE_NO_SPACES/g" {} +
+
+# search and replace files <...>
+
+echo "Package and package name within files has been renamed, Replacing Strings.xml"
```
Comment (sdsantos) [inline newproject.sh:142]: `Replacing Strings.xml` -> `replacing strings.xml`

File: newproject.sh [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    die "Invalid Package Name: $packagename (needs to follow standard pattern {com.example.package})"
+fi
+
+# Enforce title 
+if [[ "$title" =~ \ |\' ]]; then
+    die "Invalid Title: $title (can't have any spaces)"
+fi
+
+
+TITLE_NO_SPACES="${title// /}"
```
Comment (sdsantos) [inline newproject.sh:81]: I thought it couldn't have spaces already by now?

File: newproject.sh [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+# Enforce package name
+regex='^[a-z][a-z0-9_]*(\.[a-z0-9_]+)+[0-9a-z_]$'
+if ! [[ $packagename =~ $regex ]]; then
+    die "Invalid Package Name: $packagename (needs to follow standard pattern {com.example.package})"
+fi
+
+# Enforce title 
+if [[ "$title" =~ \ |\' ]]; then
+    die "Invalid Title: $title (can't have any spaces)"
+fi
```
Comment (sdsantos) [inline newproject.sh:78]: The title should also start with Uppercase, by I'm not sure if it's a strict requirement.

**Review by sdsantos — CHANGES_REQUESTED** [review]

File: newproject.sh (hunk 7 above)
Comment (Filmaluco) [reply newproject.sh:78]: If I am not mistaken its not a strict requirement

File: newproject.sh (hunk 6 above)
Comment (Filmaluco) [reply newproject.sh:81]: it was the previous system, I left as double measure :D

File: newproject.sh (hunk 7 above)
Comment (Filmaluco) [reply newproject.sh:78]: just tested, it's not

File: README.md [hunk 8]
```diff
@@ -35,28 +35,27 @@
 ### Script 
 1. Download this repository extract and open the folder
 2. Run `newproject.sh` script to create a copy with the name you want 
-3. On `app`, change the package on the `mode` lazy var
+    * Example: `./newproject.sh -p com.example.project -t project`
```
Comment (sdsantos) [inline README.md:38]: Even if it's not a requirement, it's a good practice for the title to start with Uppercase.

**Review by sdsantos — CHANGES_REQUESTED** [review]

**Pull request merged**